### This project is part of "Pandas and NumPy Fundamentals" course on DataQuest.io.
### The dataset is of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset has been downloaded from [kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) but it has been modified for this project.
Features in the dataset include:
 - dateCrawled - When this ad was first crawled. All field-values are taken from this date.
 - name - Name of the car.
 - seller - Whether the seller is private or a dealer.
 - offerType - The type of listing
 - price - The price on the ad to sell the car.
 - abtest - Whether the listing is included in an A/B test.
 - vehicleType - The vehicle Type.
 - yearOfRegistration - The year in which the car was first registered.
 - gearbox - The transmission type.
 - powerPS - The power of the car in PS.
 - model - The car model name.
 - kilometer - How many kilometers the car has driven.
 - monthOfRegistration - The month in which the car was first registered.
 - fuelType - What type of fuel the car uses.
 - brand - The brand of the car.
 - notRepairedDamage - If the car has a damage which is not yet repaired.
 - dateCreated - The date on which the eBay listing was created.
 - nrOfPictures - The number of pictures in the ad.
 - postalCode - The postal code for the location of the vehicle.
 - lastSeenOnline - When the crawler saw this ad last online.


In [129]:
import numpy as np
import pandas as pd

In [130]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

In [131]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [132]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [133]:
autos.isnull().sum()

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

 - Dataset consists of 5000 rows and 20 columns where 5 columns are of integer type and rest are of object type.
 - If we look at the data, three columns are date time but are read as Object, also price is read as Object because of \$ symbol. 
 - Only five columns have missing values
 - Column names are listed in camelcase instead of snakecase.

In [134]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [135]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

#### I have renamed the columns by converting camelcase names to snakecase and have also modified few column names to make them short

In [136]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [137]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

#### Let's first fix the price column by 
 - removing the dollar sign and ,
 - changing the data type of column to float
 - renaming the column to indicate that prices are given in dollar


In [138]:
autos['price'] = autos['price'].str.replace("$","").str.replace(",","").astype(float)
autos.rename({'price':'price_$'}, axis=1, inplace=True)

In [139]:
autos['price_$'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price_$, dtype: float64

#### we can see that all the above operations have been successfully performed. Now we can see the mean and median prices, which we couldn't see earlier due to the datatype of the price column

In [140]:
autos['price_$'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

#### we see that minimum price is 0 wich seems unrealistics, so we explore this column more to identify and remove outliers

In [141]:
autos['price_$'].value_counts()

0.0           1421
500.0          781
1500.0         734
2500.0         643
1200.0         639
1000.0         639
600.0          531
800.0          498
3500.0         498
2000.0         460
999.0          434
750.0          433
900.0          420
650.0          419
850.0          410
700.0          395
4500.0         394
300.0          384
2200.0         382
950.0          379
1100.0         376
1300.0         371
3000.0         365
550.0          356
1800.0         355
5500.0         340
1250.0         335
350.0          335
1600.0         327
1999.0         322
              ... 
2225.0           1
69997.0          1
139997.0         1
69999.0          1
4780.0           1
8930.0           1
21599.0          1
15911.0          1
10000000.0       1
5180.0           1
919.0            1
1247.0           1
5998.0           1
27020.0          1
21888.0          1
46500.0          1
2001.0           1
2459.0           1
345000.0         1
34940.0          1
2785.0           1
5248.0      

#### 1421 instances have price=0, also one instance has price=10000000 which seems too high for a used car.

#### Another column odometer, also has distance measurements, is of Object datatype. We perform following operations on odometer:
 - remove the km and comma from column values
 - change the datatype of column to float
 - rename the column to odometer_km to indicate the unit of values


In [142]:
autos['odometer'] = autos['odometer'].str.replace("km","").str.replace(",","").astype(float)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

#### all date time columns are of string datatype, we convert them to datetime datatype

In [143]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
autos['last_seen'] = pd.to_datetime(autos['last_seen'])
autos.dtypes

date_crawled          datetime64[ns]
name                          object
seller                        object
offer_type                    object
price_$                      float64
ab_test                       object
vehicle_type                  object
registration_year              int64
gear_box                      object
power_ps                       int64
model                         object
odometer_km                  float64
registration_month             int64
fuel_type                     object
brand                         object
unrepaired_damage             object
ad_created            datetime64[ns]
nr_of_pictures                 int64
postal_code                    int64
last_seen             datetime64[ns]
dtype: object

#### we can easily explore these date time columns, lets see which brands were put on sale in 4th month of 2016 and also count of each brand cars

In [144]:
autos.loc[autos['ad_created'] == '2016-04', 'brand'].value_counts()

volkswagen        343
bmw               190
opel              187
mercedes_benz     160
audi              158
ford              126
renault            71
peugeot            50
fiat               41
seat               35
smart              27
toyota             27
sonstige_autos     27
nissan             25
citroen            23
skoda              23
mazda              20
porsche            15
alfa_romeo         14
mini               14
hyundai            12
mitsubishi         12
honda              11
jeep               10
chevrolet          10
suzuki             10
kia                 9
volvo               8
dacia               8
chrysler            5
daihatsu            5
trabant             3
subaru              3
lada                2
jaguar              2
daewoo              1
lancia              1
saab                1
rover               1
Name: brand, dtype: int64

In [145]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

#### we can see minimum year is long before car was invented. also 9999 is many years into future. since car was invented in 1886 and was sold to public many years later, it would be safe to remove dates before 1900 and after 2016, because a car can't be first registered after the listing was seen.

In [146]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize=True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
2006    0.056384
2001    0.056280
2002    0.052740
1998    0.051074
2007    0.047972
2008    0.046452
2009    0.043683
1997    0.042225
2011    0.034022
2010    0.033251
1996    0.030066
2012    0.027546
2016    0.027401
1995    0.027338
2013    0.016782
2014    0.013867
1994    0.013742
1993    0.009265
2015    0.008308
1990    0.008224
1992    0.008141
1991    0.007412
1989    0.003769
1988    0.002957
1985    0.002186
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

In [147]:
autos['registration_month'].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1])

#### since 0 is not a valid month so we remove rows with 0 value of registration month

In [148]:
autos = autos[~(autos['registration_month'] == 0)]
autos['registration_month'].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  9, 11,  5,  2,  1])

### Data aggregation

In [149]:
brands_count = autos.groupby('brand').count().sort_values('date_crawled', ascending=False)['date_crawled']
brands_count.iloc[:5]

brand
volkswagen       9062
bmw              4874
opel             4538
mercedes_benz    4309
audi             3822
Name: date_crawled, dtype: int64

#### we can see that top five brands in terms of number of cars are: volkswagen, bmw, opel, mercedes_benz, audi.

#### Now we will see the brands with highest mean price

In [150]:
brands_mean_price = autos.groupby('brand').mean()['price_$'].sort_values(ascending=False)
brands_mean_price

brand
sonstige_autos    49574.985612
citroen           49367.853333
porsche           46796.678700
mercedes_benz     31976.597122
land_rover        19685.956989
jeep              12049.049505
jaguar            11925.314286
mini              10664.569652
audi               9604.788069
bmw                8734.987485
ford               7970.843361
chevrolet          6753.270161
skoda              6542.939726
dacia              5967.457627
kia                5949.240854
volkswagen         5859.364268
opel               5813.696342
hyundai            5550.663636
toyota             5227.741710
volvo              5008.017327
nissan             4895.949102
seat               4494.374372
honda              4289.222874
mazda              4269.382126
subaru             4257.516854
alfa_romeo         4205.515789
suzuki             4195.444444
smart              3676.553226
mitsubishi         3563.730435
chrysler           3373.943038
saab               3329.410959
peugeot            3184.098069
la

In [151]:
brands_mean_price.shape

(40,)

#### 'sonstige_autos' has highest mean price among 40 brands. we can also see the mean price of top 5 most frequent brands

In [152]:
brands_mean_price.loc[['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi']]

brand
volkswagen        5859.364268
bmw               8734.987485
opel              5813.696342
mercedes_benz    31976.597122
audi              9604.788069
Name: price_$, dtype: float64

#### we can also save mean prices for each brand as dictionary

In [153]:
brands_mean_price.to_dict()

{'alfa_romeo': 4205.515789473684,
 'audi': 9604.788069073784,
 'bmw': 8734.987484612227,
 'chevrolet': 6753.270161290323,
 'chrysler': 3373.9430379746836,
 'citroen': 49367.85333333333,
 'dacia': 5967.457627118644,
 'daewoo': 1072.0,
 'daihatsu': 1693.5673076923076,
 'fiat': 2892.084914182475,
 'ford': 7970.843361344538,
 'honda': 4289.222873900293,
 'hyundai': 5550.663636363636,
 'jaguar': 11925.314285714287,
 'jeep': 12049.049504950495,
 'kia': 5949.240853658536,
 'lada': 2599.75,
 'lancia': 3112.4772727272725,
 'land_rover': 19685.956989247312,
 'mazda': 4269.382126348228,
 'mercedes_benz': 31976.59712230216,
 'mini': 10664.569651741294,
 'mitsubishi': 3563.730434782609,
 'nissan': 4895.949101796407,
 'opel': 5813.6963420008815,
 'peugeot': 3184.0980694980694,
 'porsche': 46796.67870036101,
 'renault': 2548.2348894348893,
 'rover': 1677.0526315789473,
 'saab': 3329.4109589041095,
 'seat': 4494.374371859297,
 'skoda': 6542.939726027397,
 'smart': 3676.553225806452,
 'sonstige_autos':

In [154]:
brands = pd.concat([brands_count, brands_mean_price], axis=1)
brands.rename({'date_crawled':'count', 'price_$':'mean_price'}, axis=1, inplace=True)
brands.sort_values('mean_price', ascending=False)

count    mean_price
sonstige_autos    417  49574.985612
citroen           600  49367.853333
porsche           277  46796.678700
mercedes_benz    4309  31976.597122
land_rover         93  19685.956989
jeep              101  12049.049505
jaguar             70  11925.314286
mini              402  10664.569652
audi             3822   9604.788069
bmw              4874   8734.987485
ford             2975   7970.843361
chevrolet         248   6753.270161
skoda             730   6542.939726
dacia             118   5967.457627
kia               328   5949.240854
volkswagen       9062   5859.364268
opel             4538   5813.696342
hyundai           440   5550.663636
toyota            573   5227.741710
volvo             404   5008.017327
nissan            668   4895.949102
seat              796   4494.374372
honda             341   4289.222874
mazda             649   4269.382126
subaru             89   4257.516854
alfa_romeo        285   4205.515789
suzuki            261   4195.444444
smart             620   3676.553226
mitsubishi        345   3563.730435
chrysler          158   3373.943038
saab               73   3329.410959
peugeot          1295   3184.098069
lancia             44   3112.477273
fiat             1107   2892.084914
lada               24   2599.750000
renault          2035   2548.234889
trabant            44   2237.931818
daihatsu          104   1693.567308
rover              57   1677.052632
daewoo             65   1072.000000

In [155]:
print("min count:", brands['count'].min())
print("mean count:", brands['count'].mean())
print("max count:", brands['count'].max())
print("min price:", brands['mean_price'].min())
print("mean price:", brands['mean_price'].mean())
print("max price:", brands['mean_price'].max())

min count: 24
mean count: 1086.025
max count: 9062
min price: 1072.0
mean price: 9414.806397214299
max price: 49574.98561151079


In [156]:
brands.loc[brands['mean_price'] > brands['mean_price'].mean(), 'count']

audi              3822
citroen            600
jaguar              70
jeep               101
land_rover          93
mercedes_benz     4309
mini               402
porsche            277
sonstige_autos     417
Name: count, dtype: int64

#### we see that all cars with price greater than mean price have  count less than mean count except mercedes_benz and audi, which means that people are less likely to put expensive cars on sale or it could mean that more epensive cars are less sold and hence less used cars are put on sale.

#### Now we want to see if mileage has any impact on price

In [157]:
brands_mileage = autos.groupby('brand').mean()['odometer_km']
brands = pd.concat([brands, brands_mileage], axis=1)
brands.rename({'odometer_km':'mean_mileage'}, axis=1, inplace=True)

In [158]:
brands.sort_values('mean_price', ascending=False)

count    mean_price   mean_mileage
brand                                             
sonstige_autos    417  49574.985612   90155.875300
citroen           600  49367.853333  118516.666667
porsche           277  46796.678700   98537.906137
mercedes_benz    4309  31976.597122  130971.223022
land_rover         93  19685.956989  118924.731183
jeep              101  12049.049505  125990.099010
jaguar             70  11925.314286  120785.714286
mini              402  10664.569652   87972.636816
audi             3822   9604.788069  128784.667713
bmw              4874   8734.987485  132541.034058
ford             2975   7970.843361  123983.193277
chevrolet         248   6753.270161   99677.419355
skoda             730   6542.939726  110157.534247
dacia             118   5967.457627   83983.050847
kia               328   5949.240854  111646.341463
volkswagen       9062   5859.364268  128343.632752
opel             4538   5813.696342  128884.971353
hyundai           440   5550.663636  105613.636364
toyota            573   5227.741710  115191.972077
volvo             404   5008.017327  138824.257426
nissan            668   4895.949102  117485.029940
seat              796   4494.374372  121017.587940
honda             341   4289.222874  122375.366569
mazda             649   4269.382126  123543.913713
subaru             89   4257.516854  124438.202247
alfa_romeo        285   4205.515789  131578.947368
suzuki            261   4195.444444  109252.873563
smart             620   3676.553226   99104.838710
mitsubishi        345   3563.730435  125898.550725
chrysler          158   3373.943038  133797.468354
saab               73   3329.410959  144109.589041
peugeot          1295   3184.098069  126216.216216
lancia             44   3112.477273  124090.909091
fiat             1107   2892.084914  116038.843722
lada               24   2599.750000   91666.666667
renault          2035   2548.234889  127872.235872
trabant            44   2237.931818   50454.545455
daihatsu          104   1693.567308  116009.615385
rover              57   1677.052632  136140.350877
daewoo             65   1072.000000  122923.076923

#### we can see that brand "sonstige_autos" with highest mean price has very low mean mileage as compared to brand "citroen" with second highest mean price. we can safely say that if both brands had same mileage then citroen would have much higher mean price than sonstige_autos.

In [160]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

#### we see that it has german words, we translate them to english and replace them with english translation

In [161]:
german_to_eng = {
    "ja" : "yes",
    "nein" : "no"
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(german_to_eng)
autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)